src/backend/tests/unit/services/tracing/test_tracing_service.py

```diff
@@ -0,0 +1,482 @@
+        project_name: str,
+        trace_id: uuid.UUID,
+        user_id: str | None = None,
+        session_id: str | None = None,
+    ) -> None:
+        self.trace_name = trace_name
+        self.trace_type = trace_type
+        self.project_name = project_name
+        self.trace_id = trace_id
+        self.user_id = user_id
+        self.session_id = session_id
+        self._ready = True
+        self.end_called = False
+        self.get_langchain_callback_called = False
+        self.add_trace_list = []
+        self.end_trace_list = []
+
+    @property
+    def ready(self) -> bool:
+        return self._ready
+
+    def add_trace(
+        self,
+        trace_id: str,
+        trace_name: str,
+        trace_type: str,
+        inputs: dict[str, any],
+        metadata: dict[str, any] | None = None,
+        vertex=None,
+    ) -> None:
+        self.add_trace_list.append({
+            "trace_id": trace_id,
+            "trace_name": trace_name,
+            "trace_type": trace_type,
+            "inputs": inputs,
+            "metadata": metadata,
+            "vertex": vertex
+        })
+
+    def end_trace(
+        self,
+        trace_id: str,
+        trace_name: str,
+        outputs: dict[str, any] | None = None,
+        error: Exception | None = None,
+        logs=(),
+    ) -> None:
+        self.end_trace_list.append({
+            "trace_id": trace_id,
+            "trace_name": trace_name,
+            "outputs": outputs,
+            "error": error,
+            "logs": logs
+        })
+
+    def end(
+        self,
+        inputs: dict[str, any],
+        outputs: dict[str, any],
+        error: Exception | None = None,
+        metadata: dict[str, any] | None = None,
+    ) -> None:
+        self.end_called = True
+        self.inputs_param = inputs
+        self.outputs_param = outputs
+        self.error_param = error
+        self.metadata_param = metadata
+
+    def get_langchain_callback(self):
+        self.get_langchain_callback_called = True
+        return MagicMock()
+
+
+@pytest.fixture
+def mock_settings_service():
+    settings = Settings()
+    settings.deactivate_tracing = False
+    return SettingsService(settings, MagicMock())
+
+
+@pytest.fixture
+def tracing_service(mock_settings_service):
+    return TracingService(mock_settings_service)
+
+
+@pytest.fixture
+def mock_component():
+    component = MagicMock()
+    component._vertex = MagicMock()
+    component._vertex.id = "test_vertex_id"
+    component.trace_type = "test_trace_type"
+    return component
+
+
+@pytest.mark.asyncio
+async def test_start_end_tracers(tracing_service):
+    """Test starting and ending tracers."""
+    run_id = uuid.uuid4()
+    run_name = "test_run"
+    user_id = "test_user"
+    session_id = "test_session"
+    project_name = "test_project"
+    outputs = {"output_key": "output_value"}
+
+    with patch("langflow.services.tracing.service._get_langsmith_tracer", return_value=MockTracer), \
```

Comment:
I'm unable to mark the other lines through GitHub's interface.
But, this entire block needs to be indented as follows:
```suggestion
    with (
        patch("langflow.services.tracing.service._get_langsmith_tracer", return_value=MockTracer),
        patch("langflow.services.tracing.service._get_langwatch_tracer", return_value=MockTracer),
        patch("langflow.services.tracing.service._get_langfuse_tracer", return_value=MockTracer),
        patch("langflow.services.tracing.service._get_arize_phoenix_tracer", return_value=MockTracer),
    ):
```

Reply:
thanks, im gonna to fix locally. btw, is there any code style guidline for this project , I have already ran the ruff check which doesn't complain about this newline style.

Reply:
Unfortunately, our CI and the locally run Ruff are not fully aligned. While make format helps to some extent, the CI has the final say. All the formatting corrections I'm suggesting are based on the CI output I reviewed here in GitHub's interface.

Reply:
![Screenshot 2025-03-12 at 11 17 21](https://github.com/user-attachments/assets/90cc8cf7-52b5-41ff-a8c3-6fa34023eb13)

Reply:
![Screenshot 2025-03-12 at 11 18 26](https://github.com/user-attachments/assets/0632a291-8e9a-4916-95a5-7165b7a99871)
In the CI output, it provides a Git diff that you can copy and apply locally using git apply to fix everything.